Web notifications

Zaydme opened the pull request:
### What does this pull request aim to achieve? 💯
Web notifications

_Please be concise and link any related issue(s):_
closes #1637 
closes #1531 

<!--_Please do **NOT** submit this PR if you have not done the following:_-->

- [x] I have checked that no one else is working on similar changes.
- [x] I have read the latest [**README**](README.md) and understand the development workflow.
- [x] The name of this branch is: **`feature/<branch_name>`**, or **`hotfix/<branch_name>`** if this is a hotfix, where `<branch_name>` briefly describes the issue(s) resolved and is prefixed with the issue number(s) (e.g. `1127-1130-update-git-branching-model`). The branch is created in a cloned version of this repo, **not a forked repo**.
- [x] This branch is rebased/merged with the **latest staging** (or the **latest production** if it is a hotfix).
- [x] There are no merge conflicts.
- [x] There are no console warnings and errors.
- [x] On the right hand side of this PR, I have linked the appropriate issue(s) under "Issues" and added a project progress under "Projects".
- [x] The title describes the issue(s) being addressed using format: **`<issue_numbers> - <brief_issue_description>`** (e.g. `#1127/#1130 - Update Git Branching Model`)

Serhazor committed:
Rearranged Profile dropdown menu

Serhazor committed:
added sliding menu for organisations

Serhazor committed:
removed margin/lines after sign out option also after list of organisations and above the add organisation option

Serhazor committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1369

kterry1 committed:
Added private message modal

kterry1 committed:
Removed extra curly brace

kterry1 committed:
Added mobile view

kterry1 committed:
Added Mobile View for Success Message

kterry1 committed:
Added inbox page

Serhazor committed:
changed the min-width to 10rem, this should sort out the issue of the selection field to become very small at low resolution

kterry1 committed:
original post and toggle post button

kterry1 committed:
Added messaging bubles

kterry1 committed:
Fixed spacing issues with message bubbles

kterry1 committed:
Added input box border

kterry1 committed:
Redirected users to login page when not logged in

Serhazor committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1469

kterry1 committed:
Added 'No Messages' screen and mobile view for private messages

kterry1 committed:
Fixed Success Modal Button placement

kterry1 committed:
Allowed toggle between current chat and chat list in mobile view

kterry1 committed:
Disabled submit button for Modal and Inbox when no text has been typed

mannykary committed:
#1478 Microservices CI/CD for notifications backend (#1479)

- Add services folder for microservices
- Add notification backend skeleton code
- Add notification backend Terraform code
- Add notification backend Github Action workflows

Serhazor committed:
Added new component (EditNotifications) with a toggle button, passing styling to another FE endgineer

Serhazor committed:
Added new component (EditNotifications) with a toggle button, passing styling to another FE endgineer

kterry1 committed:
Added failed to send modal and cleaned up code a bit

kterry1 committed:
Redesigned Recipient's Orginal Post bubble

kterry1 committed:
Refactored Inbox components and redesigned message bubbles

kterry1 committed:
Replaced px values with respective em values

kterry1 committed:
Refactored Inbox.js for performance improvements

kterry1 committed:
Added useRef for input box focus

kterry1 committed:
Testing different comments for message modal

kterry1 committed:
Adding Notification Dropdown

kterry1 committed:
Added no more posts section

kterry1 committed:
Added overflow for extra notifications

kterry1 committed:
Improved overflow UI

kterry1 committed:
Removed redudant css fields

kterry1 committed:
Added Notification bubble and removed scrollbar

kterry1 committed:
Added Notification Mobile Icon

kterry1 committed:
Fixed scrolling of dropdown

kterry1 committed:
Moved styled components outside of functional ones

Zaydme committed:
It's an absolute mess, but you can send messages!

Zaydme committed:
messaging functionality update, ± cleanup

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
realtime users status update

Zaydme committed:
added post reference if message sent from feed

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
turn urls into hyperlinks in messages, + user status bug fix

Zaydme committed:
removed debugging leftover, and yes I use console.log to debug 😳

Zaydme committed:
optimized how things work

Zaydme committed:
don't emit JOIN_ROOM if already in same room

Zaydme committed:
load old messages (just prototype), ± mobile compatibility

mannykary committed:
Add Redis backend environment variables

Zaydme committed:
Merge pull request #1580 from FightPandemics/feature/1579-add-redis

#1579 Add Redis backend environment variables

Zaydme committed:
change socket url to work in review env

Zaydme committed:
changing socket url based on env

Zaydme committed:
add socket to nginx config

Zaydme committed:
messages grouping and timestamps

Zaydme committed:
made linter ignore some lines

Zaydme committed:
correctly add the unread mark when recipient is offline

Zaydme committed:
added avatar to reripient header

Zaydme committed:
fixed success modal visual bug

Zaydme committed:
updated status badge, unread messages indicator now shows numbers

Zaydme committed:
fixed mistake

Zaydme committed:
clear chatLog when leaving all rooms.

Zaydme committed:
styling to match design

Zaydme committed:
fixed user status state

Zaydme committed:
updated message modal to match design, fixed status state once again (hopefully)

Zaydme committed:
updated some icons to match design

Zaydme committed:
various changes

Zaydme committed:
header inbox icon with unread count

Zaydme committed:
removed unread count from the blue dot indicator

Zaydme committed:
show empty imbox screen if the user had no room, or no room selected

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
inbox shortcut with unread badge for mobile side Nav

Zaydme committed:
real-time message notifications

Zaydme committed:
don't auto join first room

Zaydme committed:
fixed a bug where socket leaves it's own room and stops receiving notifications

Zaydme committed:
display number of unread threads and not total number of unread messages

Zaydme committed:
updated message grouping timeframe, and timestamp format

Zaydme committed:
messages deleting

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Messages editing

Zaydme committed:
added the (edited) message indicator

Zaydme committed:
various improvements, code cleanup, and bug fixes

Zaydme committed:
eslint-disable-line react-hooks/exhaustive-deps

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
fix missing comma

Zaydme committed:
expandable textarea with shift&enter or text overflow

Zaydme committed:
mobile layout improvements

Zaydme committed:
prevent virtual keyboard from opening on mobile

Zaydme committed:
Merge branch 'feature/notification-drop-down/1531' into feature/1637-notifications-web-backend

Zaydme committed:
reposition notifications icon with the new translate globe icon

mannykary committed:
Add Redis container

Zaydme committed:
updated socket backend to work with clustering

Zaydme committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

Zaydme committed:
Added Redis installation guide to Readme

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

Zaydme committed:
'Block' and 'archive' front-end

Zaydme committed:
user blocking/unblocking backend

Zaydme committed:
archiving + messages requests!

Zaydme committed:
leave All rooms when leaving inbox page to keep receiving messages notifications

Zaydme committed:
Styling edited and deleted messages

Zaydme committed:
added message button on profile page.

Zaydme committed:
fixed a bug where user is not displayed on messages list after sending message from feed

Zaydme committed:
Various bug fixes

Zaydme committed:
visual style fixes

Zaydme committed:
fixed message modal avatar & user cannot message their own orgs

Zaydme committed:
don't scroll down when clicking "load more"

Zaydme committed:
Fixed message modal styling

Zaydme committed:
inbox style fixes

Zaydme committed:
regex infinite loop fix

ddveloper committed:
add notification preference in User model

Zaydme committed:
Fixed edit control out of screen

Zaydme committed:
JWT authentication for websocket

Zaydme committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
various bug fixes

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

Zaydme committed:
Merge branch 'feature/notifications' into feature/1637-notifications-web-backend

Zaydme committed:
better mobile message editing + characters counter

Zaydme committed:
keep user on inbox page after refresh

Zaydme committed:
update thread name on user update

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

Zaydme committed:
Merge branch 'feature/1637-notifications-web-backend' of https://github.com/FightPandemics/FightPandemics into feature/1637-notifications-web-backend

Zaydme committed:
fixed ios overlapping bug

mannykary committed:
Notification model (#1639)

mannykary committed:
Notification model (#1639)

Zaydme committed:
Header inbox icon for mobile

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

Zaydme committed:
fixed previous commit causing issue

Zaydme committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

Zaydme committed:
better header alignment

Zaydme committed:
remove duplicate lines

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

Zaydme committed:
Web notifications for like and comment

Zaydme committed:
Merge branch 'feature/notifications' into feature/1637-notifications-web-backend

Zaydme committed:
don't notify user of actions on their own posts

Zaydme committed:
Merge branch 'feature/1637-notifications-web-backend' of https://github.com/FightPandemics/FightPandemics into feature/1637-notifications-web-backend

Zaydme committed:
removed duplicate key created for messageModal and caused comment button to break.

Zaydme committed:
use middleware to check socket identification

Zaydme committed:
clear messageModal textarea on send success

Zaydme committed:
fixed some styling issues

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

Zaydme committed:
(for web notifications) send notification to all the user's browsers if many are open (part1)

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

Zaydme committed:
send notification to all the user's browsers if many are open (part2)

ddveloper committed:
Merge branch 'feature/1637-notifications-web-backend' into feature/1427-notifications-unsubscribe

ddveloper committed:
update action names in User notifyPrefs

ddveloper committed:
extract NotifyPreferenceSchema from User

ddveloper committed:
schema for updateNotifyPreference

claudiang1357 commented:
I was in a post to attempt updating comment (stay for quite some time - probably expired?), error was displayed as per screenshot -> Error message is overlapped with Help Board button.
![Error](https://user-images.githubusercontent.com/72559826/96624034-32d62e80-12da-11eb-8592-417f79e4d7ef.png)

ddveloper committed:
no need to export model

ddveloper committed:
keep notifyPreferenceSchema inside users

ddveloper committed:
env key for unsubscribe

ddveloper committed:
jwt-simple package is added for decoding

ddveloper committed:
jwt decoding function

wadehrarshpreet committed:
style: fix message modal design

wadehrarshpreet committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
#1448 - Add search bar on help board with a search button

 - Add tabs search
 - Make sure the index is created for testing
 - Preserve  and use regex not  search if location is defined
 - Keyword highlighting
 - Change visible HELP_TYPE filters depending on tab
 - Don't refetch posts if no filters were selected
 - Do not reset page if no filters were selected
 - Don't ignore stop words + set minimum keyword length
 - Add search mobile version
 - Fixed search safari + brave browsers visual glitches
 - Removed some filters on mobile from Poeple/Orgs to match desktop, and fixed some unnecessary post refetch
 - Remove default language from User/Org index
 - For full $text search
 - Don't submit search if keyword too short
 - Auto hide search bar on tablets and small screens

Zaydme commented:
Maybe we need a new field on the Notification models, for the "share" action, to include the used medium.
```js
 sharedVia: {
      enum: ["email", "facebook", "twitter", ...others],
      type: String,
 },
```
@mannykary @ddveloper

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

ddveloper committed:
update endpoints for notifyPrefs, add /unsubscribe

Zaydme committed:
push notifications

ddveloper committed:
add GET endpoint for /unsubscribe

ddveloper committed:
Merge branch 'feature/1637-notifications-web-backend' into feature/1427-notifications-unsubscribe

ddveloper committed:
Merge branch 'feature/1428' into feature/1427-notifications-unsubscribe

jessica-lau committed:
Add Fixture Files for Cypress Tests (#1674)

* added fixture files and updated sign in spec

* fixed cypress.json file

* added additional fixtures and updated tests

* added additional fixtures and updated tests

sujatha-m committed:
#1663 - Modified Quit filters button to Clear filter button (#1668)

* Modified Quit filters button to Clear filter button

* Moved the fix from en_GB.json to en_US.json

Zaydme committed:
(fix) user intercation with own post shouldn't trigger notifier

Zaydme committed:
Post share notifications

ddveloper committed:
follow the alphabeta order

ddveloper committed:
return limited information for /unsubscribe (unauthenticated endpoint)

ddveloper committed:
change 'post' to 'share'

ddveloper committed:
in case no notifyPreference when creating user, add default instance

jessica-lau committed:
Cypress e2e - Add cookie banner/logo POM files and refactor tests (#1706)

* added cookie banner and logo POM files and refactored e2e tests

* refactored additional tests

ddveloper committed:
undo non-relavant change

ddveloper committed:
add edit notification option in Profile editing

ddveloper committed:
complete notification editing page

ddveloper committed:
unsubscribe link (update done, render partially done)

wadehrarshpreet committed:
improvement: Chat UI fixes

wadehrarshpreet committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

wadehrarshpreet committed:
improvement: settings chat UI fix

ddveloper committed:
toggle off notification button when all false

ddveloper committed:
bug fix, was committed these by accident

emanuelquerty committed:
#614 - check email not verified (#1703)

* Fixes internal server error message on login with email not verified

* Makes the comment more concise

wadehrarshpreet committed:
improvement: alertbox action button UI fix

wadehrarshpreet committed:
improvement: long touch menu options chat

ddveloper committed:
unsubscribe page, link preference to UI

ddveloper committed:
these are not add by me ;( could be the formatter robot?

ddveloper committed:
add a success alert after update

Zaydme committed:
Revert '#1448 - Add search bar on help board with a search button' This reverts commit 0df102b186287aa9475d7a0c2adb213b3b92880a.

Zaydme committed:
fixed unread indicator position

Zaydme committed:
internationalization of messaging feature

Zaydme committed:
adjusted bubble border radius

ddveloper committed:
remove jwt-simple, use jsonwebtoken instead

ddveloper committed:
Merge branch 'feature/1637-notifications-web-backend' into feature/1427-notifications-unsubscribe

ddveloper committed:
update expireDate interface

Zaydme committed:
fixed typos en_US (messaging)

wadehrarshpreet committed:
fix: message ui fix left sidebar & other issue

wadehrarshpreet committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
Merge remote-tracking branch 'origin/feature/notifications' into feature/1637-notifications-web-backend

Zaydme committed:
Merge pull request #1677 from FightPandemics/feature/1427-notifications-unsubscribe

Feature/1427 notifications unsubscribe

wadehrarshpreet committed:
fix: message request issue fix

Zaydme committed:
ignored threads turn back into requests on new message

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Merge branch 'feature/message-modal/1457' into feature/1637-notifications-web-backend

Zaydme committed:
Merge branch 'feature/1637-notifications-web-backend' of https://github.com/FightPandemics/FightPandemics into feature/1637-notifications-web-backend

ddveloper committed:
update notification setting menu words

mannykary committed:
Merge branch 'feature/notifications' into feature/1637-notifications-web-backend

mannykary commented on backend/lib/models/Message.js:
```diff
@@ -8,7 +8,6 @@ const MESSAGE_STATUS_OPTIONS = ["deleted", "edited", "sent"]
 // -- Schema
 const postRefSchema = new Schema(
   {
-    author: Object,
```
@Zaydme was removing this intentional?

Zaydme replied:
Yes, was unnecessary, since the post author is the same as the thread recipient, and it's not really needed anyway.

mannykary commented on backend/lib/models/Message.js:
```diff
@@ -27,7 +26,8 @@ const postRefSchema = new Schema(
       type: String,
-    }
+    },
+    createdAt: Date,
```
@Zaydme Isn't this field included by default?

Zaydme replied:
No, I don't need the document creation date, I need the referenced post creation date.

mannykary commented on backend/lib/endpoints/users.js:
```diff
@@ -438,6 +478,74 @@ async function routes(app) {
+    async (req) => {
+      const { headers, body } = req;
+      
+      let decoded = {};
```
Minor: more DRY if we refactor decoding the unsubscribe token to another function since it's used in multiple places.

mannykary commented on backend/lib/models/Thread.js:
```diff
@@ -39,22 +47,14 @@ const threadSchema = new Schema(
 
 // -- Indexes
 /* eslint-disable */
 // Index to ensure unique threads between users
```
Minor: Comment should be removed since there is no longer a unique constraint

mannykary commented on backend/lib/plugins/notifier.js:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  async notify(action, post, triggeredById, sharedVia) {
+    if (!["like", "comment", "share"].includes(action)) return this.app.log.error(new Error('Invalid Notification action'));
```
Minor: If it's possible, we should try to get this from the Notification schema since that is an enum and also determines what actions are valid. It also prevents the issue where if we add a new action to the enum, we forget to also add it here.

Zaydme replied:
Good point

mannykary commented on backend/lib/plugins/notifier.js:
```diff
@@ -0,0 +1,54 @@
+
+  async notify(action, post, triggeredById, sharedVia) {
+    if (!["like", "comment", "share"].includes(action)) return this.app.log.error(new Error('Invalid Notification action'));
+    if (post.author.id.toString() == triggeredById.toString()) return; // user interacted with their own post
```
Minor: this should probably be the very first check

mannykary commented on backend/lib/plugins/notifier.js:
```diff
@@ -0,0 +1,54 @@
+    if (err) return this.app.log.error(err, 'Failed saving new Notification');
+
+    // send real-time web notification if online
+    let userIsOnline = await getSocketIdByUserId(this.app, post.author.id.toString())
```
Minor: Is there a reason why `userIsOnline` is not a `const`? It is not being reassigned to another value.

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+} = require("../../config");
+
+function onSocketConnect(socket) {
+  this.log.info(`[ws] socket connected [id: ${socket.id}]`);
```
This should probably be changed to debug so we don't log too much in Papertrail in production

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+    this.io.emit("USER_STATUS_UPDATE", { id: socket.userId, status: "online" });
+    socket.on("disconnect", () => {
+      this.log.info(`[ws] socket disconnected [socketId: ${socket.id}]`);
+      for (let room in socket.rooms) {
```
Shouldn't this be using `of` instead of `in`? Unless you intended to iterate over the keys of `socket.rooms` rather than its values. 
https://stackoverflow.com/a/52854849/2444220

Zaydme replied:
Yeah because I need the room id which is the object key. To call `socket.leave(roomId)`

mannykary replied:
Got it, just checking.

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+    });
+    socket.join(userId); // to send events to all the user's browsers, if many are open.
+    res({ code: 200, data: userId });
+    this.log.info(
```
Like my other comment, we should change this to debug

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+
+  socket.on("JOIN_ROOM", async (data, res) => {
+    userId = socket.userId;
+    let userInRoom = await isUserInRoom(this, data.threadId, socket.id);
```
Minor: should be `const` since it is not being reassigned

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+  socket.on("JOIN_ROOM", async (data, res) => {
+    userId = socket.userId;
+    let userInRoom = await isUserInRoom(this, data.threadId, socket.id);
+    var threadErr, thread;
```
Minor: should use `let` instead of `var`

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+      );
+    } else {
+      // user called joinRoom({}) inside leaveAllRooms()
+      for (let room in socket.rooms) {
```
Similar to my previous comment about using `of` instead of `in`. I'm guessing that you did intend to iterate of the keys and this is all working as expected but just wanted to make sure.

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+        return res({ code: 500, message: "Internal server error" });
+      if (!sender || !receiver)
+        return res({ code: 404, message: "User not found" });
+      let newThread = { participants: [] };
+      [sender, receiver].forEach((participant) => {
+        newThread.participants.push({
+          id: participant._id,
+          lastAccess: participant == sender ? new Date() : null,
+          name: participant.name,
+          newMessages: participant == receiver ? 1 : 0,
+          photo: participant.photo,
+          status: participant == sender ? "accepted" : "pending",
+          type: participant.type,
+        });
+      });
```
Minor: could use map:
```suggestion
      const newThread = { 
        participants: [sender, receiver].map((participant) => ({
          id: participant._id,
          lastAccess: participant == sender ? new Date() : null,
          name: participant.name,
          newMessages: participant == receiver ? 1 : 0,
          photo: participant.photo,
          status: participant == sender ? "accepted" : "pending",
          type: participant.type,
        }) 
      };
```

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+      // if user is online, force add this room to their rooms list.
+      const recipientSocketId = await getSocketIdByUserId(
+        this,
+        thread.participants.find((p) => p.id != userId).id,
```
Minor: I don't think it's a big deal right now since there is only a sender and receiver, and I'm not sure if we will have a full-blown chat with multiple users, but it might be better to have `thread.participants` be an object rather than an array for O(1) lookup. Again, probably not going to matter since we only have two participants.

ddveloper reviewed (commented):
LGTM (I still need to improve my JS skills to feel confident in reviews)

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+        { new: true, arrayFilters: [{ "userToUpdate.id": userId }] },
+      ),
+    );
+    let threadWithLastMessage = thread.toObject();
```
Minor: use `const` instead of `let` since it is not being reassigned.

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+    if (!userInRoom) {
+      // user not already in that room
+      // leave any other room, EXCEPT the unique socket room (socket.id) and unique user room (userId)
+      for (let room in socket.rooms) {
```
Similar to previous comment about using `in` instead of `of`

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+  });
+
+  socket.on("GET_USER_THREADS", async (data, res) => {
+    userId = socket.userId;
```
```suggestion
    const userId = socket.userId;
```

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+    if (threadsErr) return res({ code: 500, message: "Internal server error" });
+
+    const threadsRich = []; // Rich because will have lastMessage, and topic
+    for (let i = 0; i < threads.length; i++) {
+      const thread = threads[i].toObject();
```
```suggestion
    for (let thread of threads) {
      thread = thread.toObject();
```

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+  });
+
+  socket.on("GET_CHAT_LOG", async (data, res) => {
+    let userInRoom = await isUserInRoom(this, data.threadId, socket.id);
```
```suggestion
    const userInRoom = await isUserInRoom(this, data.threadId, socket.id);
```

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+  });
+
+  socket.on("GET_CHAT_LOG_MORE", async (data, res) => {
+    let userInRoom = await isUserInRoom(this, data.threadId, socket.id);
```
```suggestion
    const userInRoom = await isUserInRoom(this, data.threadId, socket.id);
```

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+  });
+
+  socket.on("SEND_MESSAGE", async (data, res) => {
+    userId = socket.userId;
```
```suggestion
    const userId = socket.userId;
```

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+
+  socket.on("SEND_MESSAGE", async (data, res) => {
+    userId = socket.userId;
+    let userInRoom = await isUserInRoom(this, data.threadId, socket.id);
```
```suggestion
    const userInRoom = await isUserInRoom(this, data.threadId, socket.id);
```

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+    let userInRoom = await isUserInRoom(this, data.threadId, socket.id);
+    if (!data.threadId && !userInRoom)
+      return res({ code: 401, message: "Unauthorized" });
+    var [threadErr, thread] = await this.to(
```
```suggestion
    const [threadErr, thread] = await this.to(
```

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+    if (thread.participants.find((p) => p.status == "blocked"))
+      return res({ code: 401, message: "Unauthorized" });
+    // sender is still "pending", they can't send until they accept
+    if (
+      thread.participants.find((p) => p.id == userId && p.status == "pending")
+    )
+      return res({ code: 401, message: "Unauthorized" });
```
I know this is valid syntax but the formatting looks awkward, so I would add the braces. Also `===` rather than `==` since type coercion is janky 😄 
```suggestion
    if (
      thread.participants.find((p) => p.id === userId && p.status === "pending")
    ) {
      return res({ code: 401, message: "Unauthorized" });
    }
```

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+    if (!thread) return res({ code: 401, message: "Unauthorized" });
+
+    // someone blocked someone else, no one can send a message
+    if (thread.participants.find((p) => p.status == "blocked"))
```
```suggestion
    if (thread.participants.find((p) => p.status === "blocked"))
```

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+    )
+      return res({ code: 401, message: "Unauthorized" });
+
+    let newMessage = {
```
I'm going to stop commenting on minor syntax issues that can be caught with linting and I would recommend that ESLint be run on this file to fix lint errors.

Zaydme commented:
### Important note
This branch have most/all off the messaging code. So we cannot merge notifications into staging unless messaging is already there.

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+
+  // used for unblock, accept, and unarchive (because it sets status to "accepted")
+  socket.on("UNBLOCK_THREAD", async (data, res) => {
+    userId = socket.userId;
```
Minor: I see a lot of similar logic between BLOCK_THREAD, ARCHIVE_THREAD, IGNORE_THREAD, and UNBLOCK_THREAD that could be refactored to be more DRY.

Zaydme replied:
Yeah I was thinking about making them one.
All of the `socket.js` plugin code is not final, multiple things need to be organized and optimized.

mannykary replied:
Got it, I assume you will do that in your other PR so feel free to disregard all my comments in this file for now. I don't think this should block us from merging to `feature/notifications`. You can address them in your other PR.

mannykary commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,555 @@
+      cb();
+    };
+
+    app.log.info(`[ws] websocket ready`);
```
Change to debug

mannykary commented on client/nginx/default.conf.template:
```diff
@@ -43,4 +43,16 @@ server {
     proxy_pass $s4a_domain$request_uri;
   }
+  
+   location /socket.io {
```
```suggestion
  location /socket.io {
```

mannykary commented on client/src/components/Feed/MessagesModal/MessageModal.js:
```diff
@@ -0,0 +1,194 @@
+      <RefPost>
+        <header>
+          <div className="author">
+            {t("messaging.to")}:<TextAvatar src={avatar}>{Initials}</TextAvatar> {postAuthorName}
```
Avatar image feature is now in production. Can this be changed to use the new component? See this [commit](https://github.com/FightPandemics/FightPandemics/commit/a431c3c18d70a13ee3eeaa0b1b4cf2ce1bbf2d8a) for an example and reach out to @ashminbhandari if you have specific implementation questions.

mannykary commented on client/src/components/Inbox/ChatList.js:
```diff
@@ -0,0 +1,204 @@
+            }}
+          >
+            <Badge>
+              <TextAvatar src={getReceiver(_room.participants).photo}>
```
As mentioned in another comment, we should use the actual avatar image.

mannykary commented on client/src/components/Inbox/ChatList.js:
```diff
@@ -0,0 +1,204 @@
+                <UserName>{getReceiver(_room.participants).name}</UserName>
+                <h5>
+                  {_room.lastMessage
+                    ? moment(_room.lastMessage.createdAt).format("MMM. DD")
```
The date should be localized based on the user's locale preference.

mannykary commented on client/src/components/Inbox/InputBox.js:
```diff
@@ -0,0 +1,315 @@
+    }
+  }, [inputExpanded, scrollToBottom]);
+
+  const handleSendMgessage = async () => {
```
```suggestion
  const handleSendMessage = async () => {
```

mannykary commented on client/src/components/Inbox/InputBox.js:
```diff
@@ -0,0 +1,315 @@
+
+  const handleClick = async (e) => {
+    e.preventDefault();
+    handleSendMgessage();
```
```suggestion
    handleSendMessage();
```

mannykary commented on client/src/components/Inbox/InputBox.js:
```diff
@@ -0,0 +1,315 @@
+    if (isMobile()) return;
+    if (e.key === "Enter" && !e.shiftKey) e.preventDefault();
+    if (e.key === "Enter" && !e.shiftKey && text.length > 0) {
+      handleSendMgessage();
```
```suggestion
      handleSendMessage();
```

mannykary commented on client/src/components/Inbox/RecipientHeader.js:
```diff
@@ -0,0 +1,223 @@
+            alt="Back Arrow"
+          />
+          <Badge>
+            <TextAvatar src={participant.photo}>
```
As mentioned in another comment, should use the new component

mannykary commented on client/src/components/Inbox/Settings.js:
```diff
@@ -0,0 +1,207 @@
+          .map((_room) => (
+            <ThreadContainer>
+              <Badge>
+                <TextAvatar src={getReceiver(_room.participants).photo}>
```
Use new component

mannykary commented on client/src/components/Inbox/Settings.js:
```diff
@@ -0,0 +1,207 @@
+          .map((_room) => (
+            <ThreadContainer className={"for-blocked"}>
+              <Badge>
+                <TextAvatar src={getReceiver(_room.participants).photo}>
```
Use new component

mannykary commented on client/src/components/Notifications/NotificationDropDown.js:
```diff
@@ -0,0 +1,281 @@
+}) => {
+  return (
+    <ItemContainer href={path}>
+      <TextAvatar src={avatar}>{getInitialsFromFullName(author)}</TextAvatar>
```
Use new component

mannykary commented on client/src/components/Notifications/NotificationDropDown.js:
```diff
@@ -0,0 +1,281 @@
+    dispatch({ type: "LOCAL_NOTIFICATIONS_MARK_AS_READ" });
+  };
+
+  const notificationTypes = {
```
This needs to be localized

mannykary commented on client/src/context/WebsocketContext.js:
```diff
@@ -0,0 +1,331 @@
+    }
+  }
+
+  generateNotificationText = (n) => {
```
This needs to be localized

mannykary commented on client/src/context/WebsocketContext.js:
```diff
@@ -0,0 +1,331 @@
+    if (Notification.permission === "granted") {
+      let img =
+        "https://raw.githubusercontent.com/FightPandemics/FightPandemics/master/images/fp_logo.png";
+      var notification = new Notification("You have a new notification!", {
```
This needs to be localized

mannykary commented on client/src/context/WebsocketContext.js:
```diff
@@ -0,0 +1,331 @@
+    if (!("Notification" in window)) return;
+    if (Notification.permission === "granted") {
+      let img =
+        "https://raw.githubusercontent.com/FightPandemics/FightPandemics/master/images/fp_logo.png";
```
I think we should start putting some of our assets in the Cloudfront CDN

mannykary commented on client/src/locales/translations/en_GB.json:
```diff
@@ -125,6 +125,9 @@
     "filterBy": "Filter by",
```
We should only be updating `en_US.json` since the other locales are automatically generated by Lokalise

mannykary commented on client/src/locales/translations/en_US.json:
```diff
@@ -792,5 +802,77 @@
+      "p1": "The content of the message request would apprear here", 
+      "p2": "once you select one from the message requests inbox."
+    },
+    "didBlock": "You've blocked {{username}}. unblock to receive messages from them again.",
```
```suggestion
    "didBlock": "You've blocked {{username}}. Unblock to receive messages from them again.",
```

mannykary commented on client/src/locales/translations/en_US.json:
```diff
@@ -792,5 +802,77 @@
+      "p2": "once you select one from the message requests inbox."
+    },
+    "didBlock": "You've blocked {{username}}. unblock to receive messages from them again.",
+    "wasBlocked": "You've been blocked by {{username}}. you can no longer message them.",
```
```suggestion
    "wasBlocked": "You've been blocked by {{username}}. You can no longer message them.",
```

mannykary commented on client/src/locales/translations/en_US.json:
```diff
@@ -792,5 +802,77 @@
+    },
+    "didBlock": "You've blocked {{username}}. unblock to receive messages from them again.",
+    "wasBlocked": "You've been blocked by {{username}}. you can no longer message them.",
+    "acceptPropmt": "Do you want to accept the message request? you cannot reply until you accept the request.",
```
```suggestion
    "acceptPrompt": "Do you want to accept the message request? You cannot reply until you accept the request.",
```

mannykary commented on client/src/locales/translations/en_US.json:
```diff
@@ -792,5 +802,77 @@
+    "wasBlocked": "You've been blocked by {{username}}. you can no longer message them.",
+    "acceptPropmt": "Do you want to accept the message request? you cannot reply until you accept the request.",
+    "deleteMessageQuestion": "Delete the message?",
+    "deleteMessageText": "The selected message will be permanently deleted from both yours and the recipient's devices. Are you sure you want to delete?",
```
```suggestion
    "deleteMessageText": "The selected message will be permanently deleted from both your device and the recipient's device. Are you sure you want to delete?",
```

mannykary commented on client/src/locales/translations/en_US.json:
```diff
@@ -792,5 +802,77 @@
+    "to": "To",
+    "send": "Send",
+    "done": "Done",
+    "close": "close"
```
```suggestion
    "close": "Close"
```

mannykary commented on client/src/pages/EditNotifications.js:
```diff
@@ -0,0 +1,143 @@
+            {t("profile.individual.editProfile")}
+          </CustomHeading>
+          <FillEmptySpace />
+          <ProfilePic
```
Nice, using the new avatar component here 👍🏽

mannykary commented on client/src/routes.js:
```diff
@@ -215,10 +225,22 @@ const routes = [
+    component: Inbox,
+    props: {
+      hideFooter: true,
+    },
```
Do we need to add the `loggedInOnly` prop (just like the one above for `edit-notifications`) so that it's only visible when logged in?

Zaydme replied:
No, using `loggedInOnly` causes some issues and unwanted redirects. I have other measurements in place to make it only visible to logged in.

mannykary commented on client/src/utils/relativeTime.js:
```diff
@@ -0,0 +1,34 @@
+import moment from 'moment';
+
+export default function (ISODate) {
+    moment.updateLocale("en", {
```
Do we need to handle other locales?

mannykary reviewed (commented):
This was a huge PR (I think because a lot of different branches from other PRs got merged into this one, including direct messages). Overall looks very good. A lot of the comments I have are very minor and I don't care if we address them now vs later. Any comment prefixed with "Minor:" can be addressed later.

mannykary commented:
> ### Important note
> This branch have most/all off the messaging code. So we cannot merge notifications into staging unless messaging is already there.

Is direct messages part of the MVP launch in December? If yes, I think we should be able to get both messaging and notifications merged in at the same time to staging.

Zaydme commented:
@mannykary yes it is part of MVP launch, it is almost ready I guess, we still need to add GTM stuff and do a code cleanup. Not sure if we can get e2e encryption ready before launch, but I think that's not MVP.

mannykary commented:
@Zaydme I think we can go ahead and merge into `feature/notifications` so long as we address the comments in this PR at some point. We can open another PR to do that.
